### ISSUE TITLE
use rollup instead of single-file-build

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx#1009]
Single-file-build is a hacky process relying on files being concatened and import statements being stripped. The reason for that is that it results in a single module, where classes and functions, that where exported from the source files to share them with other TS files, are no longer export to the outside world. (This greatly benefits uglyfy). So the structure is now roughly:

```typescript
/// A.ts
export function functionWithLongName() {
}

export function anotherFunctionWithLongName() {

}

/// B.ts
import {functionWithLongName} from "./A.ts"

export function() yetAnotherFunctionWithLongName {
   functionWithLongName()
}

/// mobx.ts

export {anotherFunctionWithLongName} from "./A.ts"
```
Because of the concatenation, stripping and uglification, things like `functionWithLongName` get nicely a short name in the small build. Let's check how well rollup handle this. If final build is similar; that would be great and could as well (non-minified) be used as the default module entry point.

**Comment by mweststrate** [comment]
@rossipedia interested ;-) ?

**Comment by rossipedia** [comment]
@mweststrate absolutely! I'll dig in and see what I can find out 👍

**Comment by rossipedia** [comment]
Ok I have a working build script, but with some peculiarities. It needs a little clean up, so I'll PR it tomorrow for review (it's late and I'm on 3hrs of sleep)

**Comment by rossipedia** [comment]
I also need to dig into the output to make sure the minification process works as you described. Which, it _should_ as the umd/minify steps are still the same.

**Comment by rossipedia** [comment]
Still working on this. Ran into a snag with the `.d.ts` generation and rollup, so now I'm working on a PR to `rollup-plugin-typescript2` to remedy it.

Ideally when the dust clears here we should have a fairly slim rollup-based build pipeline.

**Comment by rossipedia** [comment]
Ok, so I'm at an impasse with this at the moment.

TL;DR: definition (`*.d.ts`) generation is not really supported in the current rollup+typescript landscape.

There are two current rollup plugins for transpiling typescript during bundling

* [rollup-plugin-typescript][1] - Doesn't officially support TypeScript 2.0+
* [rollup-plugin-typescript2][2] - Doesn't generate `.d.ts` files

I thought, "Ok, so I'll just add `.d.ts` file generation to `rollup-plugin-typescript2`". Turns out, the transpilation code in that plugin _does_ generate dts files, they're just not included in the bundled output (since rollup really only cares about generated _javascript_).

So that presents a couple of options:

1. Write out the `.d.ts` structure if `declaration: true` is specified in `tsconfig.json`

This is fairly trivial to implement, but it results in a file/folder structure that mirrors the original source code structure, potentially resulting in a _lot_ of declaration files. 

2. Attempt to bundle the `.d.ts` files in the same way that rollup itself does, using TypeScript's compiler and transformation API.

Rollup itself is no help here since it's designed to only process files that are transformed into javascript, so this would have to be implemented manually (or in another package). This is where things get tricky: the current TypeScript AST transformation API is _super_ recent (v2.3+), and still very much in flux. Also, like rollup, it's focused primarily on typescript that emits javascript, and working with pure declaration files is a bit painful.

3. Bundle the `.d.ts` files manually using strings and regexs.

This is essentially the same thing that `single-file-build.js` is doing. If rolled into another project, it would have to be a bit more robust, as it would have to handle things like import/export aliases and name collisions and such. At that point we're not really gaining anything over the existing `single-file-build.js` solution.

If you still want to go with a rollup-based build, I would probably suggest option #1 and write out the entire structure to the destination folder. It's quite a bit more bytes in the package than the current `mobx.d.ts` file that gets generated, but as they're declarations only they wouldn't be included in any other package that depended on `mobx`.

So yeah... that's where I'm at on this :)


[1]: https://github.com/rollup/rollup-plugin-typescript
[2]: https://github.com/ezolenko/rollup-plugin-typescript2

**Comment by mweststrate** [comment]
I think 1 should be fine, with a bit of `.npmignore` stuff. The size of the dts files indeed doesn't matter.

Another question, with Rollup, is there no additional stuff exposed (things that where declared `export` in the source modules)? Does the `test/api.js` succeed against the rolled up bundle?

**Comment by rossipedia** [comment]
1) Only symbols that are exported by `mobx.d.ts` will be visible to code that references it, so exported stuff in nested .d.ts files isn't visible by default (at least in my testing). 

2) Running `tape test/api.js` gives me:

```
$ tape test/api.js 
TAP version 13
# correct api should be exposed
ok 1 should be equivalent
ok 2 should be equal
ok 3 should be equivalent
ok 4 should be equal

1..4
# tests 4
# pass  4

# ok
```

Which looks like it's good.

**Comment by benjamingr** [comment]
@MadaraUchina

**Comment by mweststrate** [comment]
Fixed and releases as 3.1.11 last week
